### PR TITLE
Room introduction & migration of recent search

### LINKED ISSUE #3111: Support recent search tracking with Room
Should be done:
* A kind of generic and abstract accessor to access Room data
* Secure the Room DB is created properly at start (if it is not easy to create on the fly) if necessary (first installation or migration context)
* The data structure and accessor to the recent searches (based on what exists in ObjectBox)
* Link with the existing code around the search UI (disconnect this from ObjectBox)
* Write the automated tests (Unit tests and maybe one UI test) for that new Room code (adapt what exists for ObjectBox)
* Write the migration function (for the data) from the ObjectBox recent_searches DB to the Room DB
* Write automated tests for that migration function

Once this is done #3115 will have to be implemented to secure we can test the migration.

RecentSearchEntity (Responisble for recent searches) https://github.com/kiwix/kiwix-android/blob/e2c901bf22034ea891c29ce12bc5793df7eccf52/core/src/main/java/org/kiwix/kiwixmobile/core/dao/entities/RecentSearchEntity.kt

### PR DESCRIPTION
**Pull request by gouri-panda**
Fixes #3111

### CODE REVIEW

**Comment by gouri-panda** [comment]
@MohitMaliDeveloper It's ready to review.

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliDeveloper It's ready to review.

@gouri-panda Test cases are failing, can you please fix it so I can start reviewing this PR.

**Comment by kelson42** [comment]
@gouri-panda please put correct title

File: app/src/androidTest/java/org/kiwix/kiwixmobile/RecentSearchRoomDaoTest.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    val context = ApplicationProvider.getApplicationContext<Context>()
+    database = Room.inMemoryDatabaseBuilder(context, KiwixRoomDatabase::class.java).build()
+    dao = database.recentSearchRoomDao()
+    // Save a recent search entity
+    dao.saveSearch("query 1", zimId)
+    // Search for recent search entities with a matching zimId
+    val result = dao.search(zimId).first()
+    // Verify that the result contains the saved entity
+    assertThat(result.size, equalTo(1))
+    assertThat(result[0].searchTerm, equalTo("query 1"))
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/RecentSearchRoomDaoTest.kt:58]: we are using `query 1` twice so it would be better to store in a variable and use it instead of passing a new string multiple times.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/KiwixRoomDatabase.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+@Suppress("UnnecessaryAbstractClass")
+@Database(entities = [RecentSearchRoomEntity::class], version = 1)
+abstract class KiwixRoomDatabase : RoomDatabase() {
+  abstract fun recentSearchRoomDao(): RecentSearchRoomDao
+
+  companion object {
+    private var db: KiwixRoomDatabase? = null
+    fun getInstance(context: Context): KiwixRoomDatabase {
+      val callback =
+        if (BuildConfig.BUILD_TYPE != "fdroid")
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/KiwixRoomDatabase.kt:40]: @gouri-panda This always returns true, you forgot to define the `fdroid variant` in the `core module` which we placed last time in room PR.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/remote/RoomDatabaseCallback.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    super.onCreate(db)
+    CoreApp.coreComponent.inject(this)
+    boxStore.let(::migrateRecentSearch)
+  }
+
+  fun migrateRecentSearch(boxStore: BoxStore) {
+    migrationToRoomInsert(boxStore.boxFor())
+  }
+
+  fun migrationToRoomInsert(
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/remote/RoomDatabaseCallback.kt:49]: For migrating `Recent Searches` we are using `migrateRecentSearch()` function so it would be better to make this function private.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/remote/RoomDatabaseCallback.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    box: Box<RecentSearchEntity>
+  ) {
+    val searchRoomEntityList = box.all
+    searchRoomEntityList.forEachIndexed { _, recentSearchEntity ->
+      CoroutineScope(Dispatchers.IO).launch {
+        kiwixRoomDatabase.recentSearchRoomDao()
+          .saveSearch(recentSearchEntity.searchTerm, recentSearchEntity.zimId)
+        // Todo Should we remove object store data now?
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/remote/RoomDatabaseCallback.kt:57]: Yes, that would be great to remove the user's existing data after successfully migrating data to the room.

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreActivityComponent.kt [hunk 5]
```diff
@@ -50,6 +50,7 @@ interface CoreActivityComponent {
   fun inject(historyFragment: HistoryFragment)
   fun inject(bookmarksFragment: BookmarksFragment)
   fun inject(addNoteDialog: AddNoteDialog)
+  // fun inject(roomDatabaseCallback: RoomDatabaseCallback)
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/di/components/CoreActivityComponent.kt:53]: Please remove the unused code.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@gouri-panda architecture looks good.
* But we have to use better naming for variables.
* Splitting classes of room accessor and migration code is a good approach :+1: 
* We have not done much in test cases yet, we should add the edge cases as well, as @kelson42 requested on our last room PR.

**Comment by MohitMaliFtechiz** [comment]
@gouri-panda, Please fix the `Unit coverage` failure.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt [hunk 6]
```diff
@@ -96,6 +98,7 @@ abstract class CoreMainActivity : BaseActivity(), WebViewProvider {
         exitProcess(KIWIX_INTERNAL_ERROR)
       }
     }
+    objectBoxToRoomMigrator.migrateObjectBoxDataToRoom()
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:101]: As @kelson42 said we should migrate once. This runs everytime.

File: core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/DeleteRecentSearchTest.kt [hunk 7]
```diff
@@ -19,21 +19,29 @@
...
-import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
+import org.kiwix.kiwixmobile.core.dao.RecentSearchRoomDao
 import org.kiwix.kiwixmobile.core.search.adapter.SearchListItem
 import org.kiwix.kiwixmobile.core.search.adapter.SearchListItem.RecentSearchListItem
 
 internal class DeleteRecentSearchTest {
 
   @Test
-  fun `invoke with deletes a search`() {
+  fun `invoke with deletes a search`() = runTest {
```
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/DeleteRecentSearchTest.kt]: @gouri-panda, yes I have used `runBlockingTest` but it also fails with it currently I tried with `runBlocking` let see how it's working on CI because it is always passing locally.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt (hunk 6 above)
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:101]: @gouri-panda, 
* We are migrating once because after migrating we are removing the data from `objectbox` so it won't migrate again it will check if there is any data available for migration then it will run migration otherwise it will skip the process.
* I have not placed the migration available condition here because it needs to create extra variables in `CoreMainActivity`  and needs to place `if` `else` conditions here(for recent searches and we will migrate other stuff as well in the future like NotesEntity etc) so to avoid placing extra conditions we are directly checking in `ObjectBoxToRoomMigrator` class.

Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:101]: @MohitMaliFtechiz  Ok. It's not the best approach but we can work around this. We can put a `sharedpPreferce` boolean value that checks whether it has successfully migrated previously or not. We should do this here.
>to avoid placing extra conditions we are directly checking in ObjectBoxToRoomMigrator class.

We should do all the migarations in the `ObjecBoxToMigrator` class.

Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:101]: > @MohitMaliFtechiz Ok. It's not the best approach but we can work around this. We can put a sharedpPreferce boolean value that checks whether it has successfully migrated previously or not. We should do this here.

Okay

> We should do all the migarations in the ObjecBoxToMigrator class.

Yes, we are only migrating the data in `ObjectBoxToRoomMigrator` class, migrating condition is handled by the `migrateRecentSearch` method itself.

Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt]: @gouri-panda, I have did the changes you requested.

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz @kelson42 LGTM! @MohitMaliFtechiz Thanks for the work 🎉

**Review by MohitMaliFtechiz — APPROVED** [review]
@kelson42 , @gouri-panda LGTM. @gouri-panda Thanks for your hard work :tada: :tada: .

File: app/src/androidTest/java/org/kiwix/kiwixmobile/ObjectBoxToRoomMigratorTest.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  @After
+  fun cleanup() {
+    kiwixRoomDatabase.close()
+    boxStore.close()
+  }
+
+  @Test
+  fun migrateRecentSearch_shouldInsertDataIntoRoomDatabase() = runBlocking {
+    val box = boxStore.boxFor(RecentSearchEntity::class.java)
+    val expectedSearchTerm = "test search"
```
Comment (kelson42) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/ObjectBoxToRoomMigratorTest.kt:66]: Please test edge case, this a feedback I had already given to previous PR. Please read the comments there.

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt [hunk 9]
```diff
@@ -96,6 +98,9 @@ abstract class CoreMainActivity : BaseActivity(), WebViewProvider {
         exitProcess(KIWIX_INTERNAL_ERROR)
       }
     }
+    if (!sharedPreferenceUtil.prefIsRecentSearchMigrated) {
```
Comment (kelson42) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreMainActivity.kt:101]: I kind of have the feeling that once all the element are migrated we should have only on preference. I would put a reminder to do that at the end of the coding effort.

**Review by kelson42 — CHANGES_REQUESTED** [review]
Only a few remarks in the code. All in one it looks good (but I looked at a very high level). We will start to merge this once `3.7.0` is migrated anyway.

File: app/build.gradle.kts [hunk 10]
```diff
@@ -75,7 +75,15 @@ android {
...
     }
+    create("fdroid") { // Configuration for F-Droid flavor
+      // dimension = "default"
+      initWith(getByName("release"))
+      matchingFallbacks += "release"
+      applicationIdSuffix = ".fdroid"
+      versionNameSuffix = "-fdroid"
+    }
   }
+
```
[the remark this answers is not in the document]
Comment (gouri-panda) [reply app/build.gradle.kts:86]: >fdroidImplementation(Libs.objectbox_gradle_plugin) {
    exclude(group = "io.objectbox", module = "objectbox-gradle-plugin")
  }
This code removes the deps from the fdroid build. As we're still depending on the objectbox on the other codes. We'll remove it after the end of the migration.

[the remark this answers is not in the document]
Comment (kelson42) [reply app/build.gradle.kts:86]: OK, so how the app works without them?

[the remark this answers is not in the document]
Comment (gouri-panda) [reply app/build.gradle.kts:86]: We're only excluding in the fdroid build. Other than that pretty much default for all default configurations. If we wanna build a new flavour it'll come out from default.

[the remark this answers is not in the document]
Comment (gouri-panda) [reply app/build.gradle.kts:86]: @kelson42 At this stage we can't build fully for the fdroid as we haven't migrated all the funcality. As soon we've migrated all the stuff & little more build configuration we can build successfully for the fdroid.

[the remark this answers is not in the document]
Comment (kelson42) [reply app/build.gradle.kts:86]: @gouri-panda The overall approach is to consider`develop` branch potentially ready to release any time. Therefore the rule is to not merged features which are halfway done as far as possible. Here is there is no reason to introduce now a feature which is halfway done: this is why I said "do that once at the end of the Room project".

**Comment by kelson42** [comment]
@MohitMaliFtechiz Would be good if if we could get this pr ready for a new review today

File: app/build.gradle.kts (hunk 10 above)
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/build.gradle.kts:86]: Okay, so we should remove this code from this PR because we will not publish our application on F-Droid until we successfully remove the Object Box from our codebase. F-Droid does not allow us to publish our app with the Object Box code. Additionally, we don't need this code for the Play Store build and other platform build types. Therefore, it is unused in this context. Please let me know if my explanation makes sense to you, @kelson42 and @gouri-panda.

[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply app/build.gradle.kts:78]: I have removed the fdroid variant and its dependencies from this PR.

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz @kelson42 LGTM 👍
